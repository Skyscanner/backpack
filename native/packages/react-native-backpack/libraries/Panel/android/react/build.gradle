apply plugin: 'com.android.library'
apply plugin: 'kotlin-android'

android {
  compileSdkVersion rootProject.ext.compileSdkVersion
  buildToolsVersion rootProject.ext.buildToolsVersion


  defaultConfig {
    minSdkVersion rootProject.ext.minSdkVersion
    targetSdkVersion rootProject.ext.targetSdkVersion
    versionCode 1
    versionName "1.0"
  }
}

dependencies{
  implementation "com.facebook.react:react-native:+" // From node_modules
  implementation project(':panel-native')
  implementation "org.jetbrains.kotlin:kotlin-stdlib-jdk7:$kotlin_version"
}
repositories {
  mavenCentral()
}
