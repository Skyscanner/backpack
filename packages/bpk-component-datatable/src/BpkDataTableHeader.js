/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2016 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/* @flow strict */

import React from 'react';

import { cssModules } from '../../bpk-react-utils';
import BpkSmallArrowDownIcon from '../../bpk-component-icon/sm/arrow-down';
import BpkSmallArrowUpIcon from '../../bpk-component-icon/sm/arrow-up';
import { withRtlSupport } from '../../bpk-component-icon';

import { SORT_DIRECTION_TYPES } from './sort-types';
import STYLES from './BpkDataTableHeader.module.scss';
import { type ColumnType } from './common-types';

const DownIcon = withRtlSupport(BpkSmallArrowDownIcon);
const UpIcon = withRtlSupport(BpkSmallArrowUpIcon);

const getClassName = cssModules(STYLES);

const KEYCODES = {
  ENTER: 13,
  SPACEBAR: 32,
};

const BpkDataTableHeader = ({ column }: { column: ColumnType }) => {
  const {
    defaultSortDirection,
    disableSortBy,
    flexGrow,
    headerClassName,
    headerStyle,
    isSorted,
    isSortedDesc,
    label,
    minWidth,
    sortDirection,
    width,
  } = column;

  const onHeaderClick = () => {
    if (disableSortBy === true) {
      return;
    }
    let isDescending;

    if (isSorted) {
      isDescending = isSortedDesc === false;
    } else if (sortDirection) {
      isDescending = sortDirection === SORT_DIRECTION_TYPES.DESC;
    } else if (defaultSortDirection) {
      isDescending = defaultSortDirection === SORT_DIRECTION_TYPES.DESC;
    } else {
      isDescending = false;
    }

    column.toggleSortBy(isDescending);
  };

  const handleKeyboardEvent = (event: React.KeyboardEvent<HTMLElement>) => {
    if (
      event.keyCode === KEYCODES.ENTER ||
      event.keyCode === KEYCODES.SPACEBAR
    ) {
      event.preventDefault();
      onHeaderClick();
    }
  };

  const headerClassNames = getClassName(
    'bpk-data-table-column__header',
    headerClassName,
  );
  const upIconClassNames = getClassName(
    'bpk-data-table-column__sort-icon--up',
    'bpk-data-table-column__sort-icon',
    isSorted && !isSortedDesc && 'bpk-data-table-column__sort-icon--selected',
  );

  const downIconClassNames = getClassName(
    'bpk-data-table-column__sort-icon--down',
    'bpk-data-table-column__sort-icon',
    isSorted && isSortedDesc && 'bpk-data-table-column__sort-icon--selected',
  );

  return (
    <div
      {...column.getHeaderProps({
        style: {
          width,
          minWidth,
          flexGrow,
          ...headerStyle,
        },
        className: headerClassNames,
      })}
    >
      <span
        onClick={onHeaderClick}
        onKeyDown={handleKeyboardEvent}
        aria-label={label}
        aria-pressed={isSorted}
        role="button"
        tabIndex={0}
      >
        {column.render('Header')}
      </span>
      {!disableSortBy && (
        <div
          className={getClassName('bpk-data-table-column__sort-icons')}
          key="sort"
          aria-hidden
        >
          <UpIcon
            onClick={() => column.toggleSortBy(false)}
            className={upIconClassNames}
          />
          <DownIcon
            onClick={() => column.toggleSortBy(true)}
            className={downIconClassNames}
          />
        </div>
      )}
    </div>
  );
};

export default BpkDataTableHeader;
