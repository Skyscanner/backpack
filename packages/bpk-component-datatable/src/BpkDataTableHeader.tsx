/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2016 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import type { KeyboardEvent } from 'react';

import { withRtlSupport } from '../../bpk-component-icon';
import BpkSmallArrowDownIcon from '../../bpk-component-icon/sm/arrow-down';
import BpkSmallArrowUpIcon from '../../bpk-component-icon/sm/arrow-up';
import { cssModules } from '../../bpk-react-utils';
// @ts-expect-error Untyped import. See `decisions/imports-ts-suppressions.md`.
// @ts-expect-error Untyped import. See `decisions/imports-ts-suppressions.md`.

import { SORT_DIRECTION_TYPES } from './common-types';

import STYLES from './BpkDataTableHeader.module.scss';

const DownIcon = withRtlSupport(BpkSmallArrowDownIcon);
const UpIcon = withRtlSupport(BpkSmallArrowUpIcon);

const getClassName = cssModules(STYLES);

const KEYCODES = {
  ENTER: 13,
  SPACEBAR: 32,
};


/**
 * Internal component to render the header of a column.
 * @returns {JSX.Element} data table header component
 */
const BpkDataTableHeader = ({ column }: { column: any }) => {
  const {
    defaultSortDirection,
    disableSortBy,
    flexGrow,
    headerClassName,
    headerStyle,
    isSorted,
    isSortedDesc,
    label,
    minWidth,
    sortDirection,
    width,
  } = column;

  /**
   * Callback for when the header is clicked to sort the column unless disableSortBy is true.
   * @returns {void}
   */
  const onHeaderClick = () => {
    if (disableSortBy === true) {
      return;
    }
    let isDescending;

    if (isSorted) {
      isDescending = isSortedDesc === false;
    } else if (sortDirection) {
      isDescending = sortDirection === SORT_DIRECTION_TYPES.DESC;
    } else if (defaultSortDirection) {
      isDescending = defaultSortDirection === SORT_DIRECTION_TYPES.DESC;
    } else {
      isDescending = false;
    }

    column.toggleSortBy(isDescending);
  };

  /**
   * Callback for handling keyboard events on the header to sort the column.
   * @param {KeyboardEvent} event keyboard event on the header
   * @returns {void}
   */
  const handleKeyboardEvent = (event: KeyboardEvent<HTMLElement>) => {
    if (
      event.keyCode === KEYCODES.ENTER ||
      event.keyCode === KEYCODES.SPACEBAR
    ) {
      event.preventDefault();
      onHeaderClick();
    }
  };

  const headerClassNames = getClassName(
    'bpk-data-table-column__header',
    headerClassName,
  );
  const upIconClassNames = getClassName(
    'bpk-data-table-column__sort-icon--up',
    'bpk-data-table-column__sort-icon',
    isSorted && !isSortedDesc && 'bpk-data-table-column__sort-icon--selected',
  );

  const downIconClassNames = getClassName(
    'bpk-data-table-column__sort-icon--down',
    'bpk-data-table-column__sort-icon',
    isSorted && isSortedDesc && 'bpk-data-table-column__sort-icon--selected',
  );

  return (
    <div
      {...column.getHeaderProps({
        style: {
          width,
          minWidth,
          flexGrow,
          ...headerStyle,
        },
        className: headerClassNames,
      })}
    >
      <span
        onClick={onHeaderClick}
        onKeyDown={handleKeyboardEvent}
        aria-label={label}
        aria-pressed={isSorted}
        role="button"
        tabIndex={0}
      >
        {column.render('Header')}
      </span>
      {!disableSortBy && (
        <div
          className={getClassName('bpk-data-table-column__sort-icons')}
          key="sort"
          aria-hidden
        >
          <div className={upIconClassNames}>
            <UpIcon
              onClick={() => column.toggleSortBy(false)}
            />
          </div>
          <div className={downIconClassNames}>
            <DownIcon
              onClick={() => column.toggleSortBy(true)}
            />
          </div>
        </div>
      )}
    </div>
  );
};

export default BpkDataTableHeader;
