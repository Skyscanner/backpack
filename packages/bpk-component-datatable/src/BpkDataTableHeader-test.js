// /*
//  * Backpack - Skyscanner's Design System
//  *
//  * Copyright 2016 Skyscanner Ltd
//  *
//  * Licensed under the Apache License, Version 2.0 (the "License");
//  * you may not use this file except in compliance with the License.
//  * You may obtain a copy of the License at
//  *
//  *   http://www.apache.org/licenses/LICENSE-2.0
//  *
//  * Unless required by applicable law or agreed to in writing, software
//  * distributed under the License is distributed on an "AS IS" BASIS,
//  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  * See the License for the specific language governing permissions and
//  * limitations under the License.
//  */

import { render } from '@testing-library/react';

import BpkDataTableHeader from './BpkDataTableHeader';

const columnLabel = 'Name';
const sortDirection = 'ASC';

const mockColumnData = {
  defaultSortDirection: sortDirection,
  disableSortBy: false,
  flexGrow: 0,
  headerClassName: 'my-header-class',
  headerStyle: {},
  isSorted: false,
  isSortedDesc: false,
  label: columnLabel,
  minWidth: 0,
  sortDirection,
  width: 100, // eslint-disable-line  backpack/use-tokens
  getHeaderProps: jest.fn(() => ({})),
  render: jest.fn(() => columnLabel),
};

describe('BpkDataTableHeader', () => {
  it('renders sort icons', () => {
    const { asFragment } = render(
      <BpkDataTableHeader key="name" column={mockColumnData} />,
    );
    expect(asFragment()).toMatchSnapshot();
  });
  it('does not render sort icons if sorting disabled', () => {
    const { asFragment } = render(
      <BpkDataTableHeader
        key="name"
        column={{ ...mockColumnData, disableSortBy: true }}
      />,
    );
    expect(asFragment()).toMatchSnapshot();
  });
  it('renders sort icon with down icon selected', () => {
    const { asFragment } = render(
      <BpkDataTableHeader
        key="name"
        column={{ ...mockColumnData, isSorted: true, isSortedDesc: true }}
      />,
    );
    expect(asFragment()).toMatchSnapshot();
  });
  it('renders sort icon with up icon selected', () => {
    const { asFragment } = render(
      <BpkDataTableHeader
        key="name"
        column={{ ...mockColumnData, isSorted: true }}
      />,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
