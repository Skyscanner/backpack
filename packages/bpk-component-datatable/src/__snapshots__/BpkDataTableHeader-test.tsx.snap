// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BpkDataTableHeader does not render sort icons if sorting disabled 1`] = `
<DocumentFragment>
  <div>
    <span
      aria-label="Name"
      aria-pressed="false"
      role="button"
      tabindex="0"
    >
      Name
    </span>
  </div>
</DocumentFragment>
`;

exports[`BpkDataTableHeader renders sort icon with down icon selected 1`] = `
<DocumentFragment>
  <div>
    <span
      aria-label="Name"
      aria-pressed="true"
      role="button"
      tabindex="0"
    >
      Name
    </span>
    <div
      aria-hidden="true"
      class="bpk-data-table-column__sort-icons"
    >
      <div
        class="bpk-data-table-column__sort-icon--up bpk-data-table-column__sort-icon"
      >
        <svg
          aria-hidden="true"
          class="bpk-icon--rtl-support"
          height="1rem"
          viewBox="0 0 24 24"
          width="1rem"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M7.463 15.75h9.074a1.358 1.358 0 0 0 1.11-2.251l-4.354-4.77a1.53 1.53 0 0 0-2.184-.04l-4.718 4.77a1.357 1.357 0 0 0 1.072 2.291"
          />
        </svg>
      </div>
      <div
        class="bpk-data-table-column__sort-icon--down bpk-data-table-column__sort-icon bpk-data-table-column__sort-icon--selected"
      >
        <svg
          aria-hidden="true"
          class="bpk-icon--rtl-support"
          height="1rem"
          viewBox="0 0 24 24"
          width="1rem"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M16.537 8.25H7.463a1.358 1.358 0 0 0-1.11 2.251l4.354 4.77a1.53 1.53 0 0 0 2.184.04l4.718-4.77a1.357 1.357 0 0 0-1.072-2.291"
          />
        </svg>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`BpkDataTableHeader renders sort icon with up icon selected 1`] = `
<DocumentFragment>
  <div>
    <span
      aria-label="Name"
      aria-pressed="true"
      role="button"
      tabindex="0"
    >
      Name
    </span>
    <div
      aria-hidden="true"
      class="bpk-data-table-column__sort-icons"
    >
      <div
        class="bpk-data-table-column__sort-icon--up bpk-data-table-column__sort-icon bpk-data-table-column__sort-icon--selected"
      >
        <svg
          aria-hidden="true"
          class="bpk-icon--rtl-support"
          height="1rem"
          viewBox="0 0 24 24"
          width="1rem"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M7.463 15.75h9.074a1.358 1.358 0 0 0 1.11-2.251l-4.354-4.77a1.53 1.53 0 0 0-2.184-.04l-4.718 4.77a1.357 1.357 0 0 0 1.072 2.291"
          />
        </svg>
      </div>
      <div
        class="bpk-data-table-column__sort-icon--down bpk-data-table-column__sort-icon"
      >
        <svg
          aria-hidden="true"
          class="bpk-icon--rtl-support"
          height="1rem"
          viewBox="0 0 24 24"
          width="1rem"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M16.537 8.25H7.463a1.358 1.358 0 0 0-1.11 2.251l4.354 4.77a1.53 1.53 0 0 0 2.184.04l4.718-4.77a1.357 1.357 0 0 0-1.072-2.291"
          />
        </svg>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`BpkDataTableHeader renders sort icons 1`] = `
<DocumentFragment>
  <div>
    <span
      aria-label="Name"
      aria-pressed="false"
      role="button"
      tabindex="0"
    >
      Name
    </span>
    <div
      aria-hidden="true"
      class="bpk-data-table-column__sort-icons"
    >
      <div
        class="bpk-data-table-column__sort-icon--up bpk-data-table-column__sort-icon"
      >
        <svg
          aria-hidden="true"
          class="bpk-icon--rtl-support"
          height="1rem"
          viewBox="0 0 24 24"
          width="1rem"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M7.463 15.75h9.074a1.358 1.358 0 0 0 1.11-2.251l-4.354-4.77a1.53 1.53 0 0 0-2.184-.04l-4.718 4.77a1.357 1.357 0 0 0 1.072 2.291"
          />
        </svg>
      </div>
      <div
        class="bpk-data-table-column__sort-icon--down bpk-data-table-column__sort-icon"
      >
        <svg
          aria-hidden="true"
          class="bpk-icon--rtl-support"
          height="1rem"
          viewBox="0 0 24 24"
          width="1rem"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M16.537 8.25H7.463a1.358 1.358 0 0 0-1.11 2.251l4.354 4.77a1.53 1.53 0 0 0 2.184.04l4.718-4.77a1.357 1.357 0 0 0-1.072-2.291"
          />
        </svg>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
